extension/src/documents.ts

```diff
@@ -103,14 +103,26 @@ export class DocumentManager{
 		}
 
 		// This builds the '_workspace' *.LSP Memory Document placeholder set
-		// 		This feature does make the "fair assumption" that AutoCAD machines has plenty of memory to be holding all this information
-		// 		This was stress tested with a root workspace folder containing 10mb of *.LSP files and the memory footprint increased less than 50mb
+		// 		This feature does make the "fair assumption" that AutoCAD machines have plenty of memory to be holding all this information
+		// 		The impact of creating read-only documents was stress tested with a root workspace folder containing 10mb of *.LSP files
+		//		and the memory footprint from the readonlyDocument's increased the memory (sustained) by less than 50mb		
 		vscode.workspace.findFiles("**").then((items: vscode.Uri[]) => {
 			items.forEach((fileUri: vscode.Uri) => {			
 				if (this.getSelectorType(fileUri.fsPath) === DocumentManager.Selectors.lsp) {
-					this._workspace.set(fileUri.fsPath, null); // initialized as null to blueprint the workspace files without the overhead of reading their contents
+					this._workspace.set(fileUri.fsPath, ReadonlyDocument.open(fileUri.fsPath));
 				}
 			});
+		}).then(() => {
+			AutoLispExt.Documents.WorkspaceDocuments.forEach(d => {
+				// pre-caching the AtomsForest is a more significant memory jump because of its parent/child structure
+				// However, on 10mb's of LSP's in the workspace, this is only about a 90mb sustained jump and adds a lot of performance
+				// in other areas that makes this worth the cost to passively (non-blocking) pre-load the data.
+				// Also note, generating the pre-atomsForest does cause a massive memory spike (500mb) until the workspace is fully loaded.
+				d.updateAtomsForest();
```

Comment:
In my local env the F12 can not work to switch to your branch, and I see it is busy in this call stack:
```
<anonymous> (c:\Program Files\Microsoft VS Code\resources\app\out\bootstrap-fork.js:5)
o (c:\Program Files\Microsoft VS Code\resources\app\out\bootstrap-fork.js:5)
<anonymous> (c:\Program Files\Microsoft VS Code\resources\app\out\bootstrap-fork.js:5)
tokenizeString (e:\test\AutoLispExt\extension\src\format\parser.ts:131)
<anonymous> (e:\test\AutoLispExt\extension\src\project\readOnlyDocument.ts:131)
<anonymous> (e:\test\AutoLispExt\out\project\readOnlyDocument.js:8)
<anonymous> (e:\test\AutoLispExt\out\project\readOnlyDocument.js:4)
updateAtomsForest (e:\test\AutoLispExt\out\project\readOnlyDocument.js:117)
<anonymous> (e:\test\AutoLispExt\extension\src\documents.ts:121)
<anonymous> (e:\test\AutoLispExt\extension\src\documents.ts:116)
processTicksAndRejections (Unknown Source:94)
Promise.then (Unknown Source:1)
initialize (e:\test\AutoLispExt\extension\src\documents.ts:115)
DocumentManager (e:\test\AutoLispExt\extension\src\documents.ts:259)
initialize (e:\test\AutoLispExt\extension\src\context.ts:39)
activate (e:\test\AutoLispExt\extension\src\extension.ts:25)
_callActivateOptional (c:\Program Files\Microsoft VS Code\resources\app\out\vs\workbench\services\extensions\node\extensionHostProcess.js:933)
_callActivate (c:\Program Files\Microsoft VS Code\resources\app\out\vs\workbench\services\extensions\node\extensionHostProcess.js:933)
<anonymous> (c:\Program Files\Microsoft VS Code\resources\app\out\vs\workbench\services\extensions\node\extensionHostProcess.js:932)
processTicksAndRejections (Unknown Source:94)
```

Reply:
Could you help me understand what you mean by busy? I am actively trying to implement the variable definition side of this. Could that be the problem?

Reply:
I am going to guess there may be a bad document in your workspace. It looks to me like the error is being generated while trying to generate the atomsForest on the ReadonlyDocument. Since I don't have this problem, could you please surround line 121 of `documents.ts` with a Try/Catch and tell me what scenario is causing this; email me the document maybe. I'll gladly fix it if I had something to go on, but looking at the updateAtomsForest function, it already has the stuff I was worried about in a Try/Catch.

Reply:
See this pics(and let me know if you need the LSP file I used):
![F12issue](https://user-images.githubusercontent.com/68938334/98617727-76f29880-22b4-11eb-885f-68f01b1bcaea.gif)

Reply:
And check your email for the lisp file

Reply:
Thanks Josh!
It looks like I hit the performence issue here. I opened a folder which contains about 30 lisp files. I will test to open a .prj file which referece 40 lisp files, and that is a real product source codes.

Reply:
I was quite concerned about the performance issues, but I thought these were passively loading in the background and shouldn't cause anything other than a blocked Project Explorer. Can you share your computer specs? I have 480 LSP files around 10mb's and the entire workspace finishes loading in (at most) 30 seconds for me. The first part that just makes the ReadonlyDocuments (you can see something in the status bar) is less than 10 seconds. We can go back to the demand-load at the first F12, but that will lock the user interface long enough that people may think they are crashing.

Note: PRJ references are demand loaded, so you could be in for a wait on those if it has many large references.

Reply:
I tested this by deleting an open parenthesis. It does break a degree of its abilities, but I think that is working as intended because it can't properly analyze/navigate the code. No noticeable crashes though.

FYI, my specs: i7-8086k, 32GB ram, 2TB SSD primary. However, the workspace folder I use is on a NAS that uses traditional spin drives for long term storage and a 0.5TB SSD's for memory resident (recently used) files. So, I am at least slowed a bit by networking.

Reply:
this is my machine configuration:
![image](https://user-images.githubusercontent.com/68938334/98631186-bd0a2500-22d1-11eb-86e0-3940d86ceac8.png)

Reply:
My processor runs at 4ghz, but I would think your dual processors would match it. So, I am not quite sure why 30 files would take any time at all. Theoretically, if I can do 480 in 30s, then I would think yours would be handling 30 in 5s or less. Its concerning because I've based the performance on a large scale thinking I was covering the bases and maybe that isn't the case. Is your workspace particularly large with non LSP's by chance? That is about the only thing I can think of that could be causing this massive disparity. Mine has a bit of other stuff, but the bulk is definitively LSP files.

Reply:
Here is an idea. Would if, we go back to demand loading the atomsForest, but do the F12 workspace (specifically) searches with a brute force `/(DEFUN|DEFUN-Q)TheirSelection/ig.test(document.fileContent.replace(/\s/g))` to detect if its even worth generating the atomsForest in the first place. That should theoretically stop the unnecessary generation of every documents atomsForest except when they legit come into scope. This should have performance implications for the PRJ's as well. Thoughts?

Reply:
I agree that it should cache there, maybe it needs to find a async way to do cache.
I test it in a real product, it look like the performence is about 5~10 seconds, and it doesn't find the one I expected:
![bigprj](https://user-images.githubusercontent.com/68938334/98675416-12185c00-230f-11eb-846a-c7070733d91c.gif)

Reply:
PRJ's are demand loaded and that hasn't been tested on the large scale your talking about. Later tonight I will create some test scenarios and see if I can replicate the problem. Probably test my pre-check idea's viability too. At the very least it should be helpful for identifying things that should have found something that ultimately didn't. **Are the LSP's inside the PRJ also in your workspace or do they just exist in the PRJ?** I'll test both, I am just curious about your situation.

I did add the async decoration to the function that initializes the atomsForest, I suppose that not being called with an await could be causing it to gloss over the data pull. Maybe that needs to be removed.

Reply:
For the second GIF I only open a prj file, and there are no workspace opened and no folder opened.
And for the function `empty_trmlst` it has function body, but F12 can not find it. This is also weird.

I am also narrowdown the issue here. And I think for a product prj file it is normal to have hundreds of source files, as the issue https://github.com/Autodesk-AutoCAD/AutoLispExt/issues/47 mentioned.

Reply:
I have now tested this with a non-workspace and PRJ. It worked for me. I also tested this with a workspace and a PRJ containing the same files, but from another folder outside of the workspace. It successfully opened the PRJ version as the result. In short, give it another try and lets see if we have a winner now.

Reply:
To get the latest codes I tried to open the large prj files, it get hang there:
```
get line (c:\Program Files\Microsoft VS Code\resources\app\out\vs\workbench\services\extensions\node\extensionHostProcess.js:513)
offsetAt (e:\test\AutoLispExt\extension\src\project\readOnlyDocument.ts:185)
validateRange (e:\test\AutoLispExt\extension\src\project\readOnlyDocument.ts:255)
getText (e:\test\AutoLispExt\extension\src\project\readOnlyDocument.ts:239)
readComments (e:\test\AutoLispExt\extension\src\format\parser.ts:62)
tokenizeString (e:\test\AutoLispExt\extension\src\format\parser.ts:99)
updateAtomsForest (e:\test\AutoLispExt\extension\src\project\readOnlyDocument.ts:131)
addLispFileNode2ProjectTree (e:\test\AutoLispExt\extension\src\project\projectTree.ts:239)
ParseProjectDocument (e:\test\AutoLispExt\extension\src\project\openProject.ts:126)
OpenProjectFile (e:\test\AutoLispExt\extension\src\project\openProject.ts:43)
<anonymous> (e:\test\AutoLispExt\extension\src\project\openProject.ts:28)
fulfilled (e:\test\AutoLispExt\out\project\openProject.js:5)
processTicksAndRejections (Unknown Source:94)
Promise.then (Unknown Source:1)
step (e:\test\AutoLispExt\out\project\openProject.js:7)
fulfilled (e:\test\AutoLispExt\out\project\openProject.js:5)
processTicksAndRejections (Unknown Source:94)
Promise.then (Unknown Source:1)
step (e:\test\AutoLispExt\out\project\openProject.js:7)
<anonymous> (e:\test\AutoLispExt\out\project\openProject.js:8)
```

Reply:
So, that error is happening while loading the atomsForest, but within code I didn't write. I can not pre-load the atomsforest to protect the integrity of the load, but I assume that won't ultimately solve the problem. I don't expect you to email me your full program, but there is clearly something in there the parser doesn't like. Please try to figure out what that is.

Reply:
I am going to guess that Line:62 of parser.ts is putting a null value into `readComments()` and that is bubbling up to the `offsetAt()`
I'd suggest adding a conditional breakpoint on line 62 !(startPos2d && endPos2d)

Reply:
I am narrowing down the issue. maybe I can take some time this weekend.
it looks like for each F12 it build the S-expression each time? it looks like `getMemoryDocument` return a new memory document.

Reply:
getMemoryDocument kind of 'casts' the active TextDocument to a ReadonlyDocument so it is in the same context as everything else. It would do this same operation for anything in the `AutoLispExt.Documents.OpenedDocuments` getter. It kind of has to if we want to insure that unsaved edits are captured. For the same reasons, the atomsForest must be regenerated. 
Anything that is not in an opened state would already have a ReadonlyDocument representation. This would include the PRJ & workspace files. The workspace files are no longer pre-loading the atomsForest, but the PRJ's are. I did this because a load operation when opening a PRJ seems like a much more explicit intent, a load time on open makes some degree of sense from a user perspective and you told me we could expect these to potentially be excessively large; which only substantiates allowing a load time from my perspective. We can shut this off, but this problem will not go away. There is very clearly a bug in the parser we need to track down.

Reply:
I test the following case:

1.  start clean vscode
2.  open 1.lsp
3.  open 2.lsp ;; in 2.lsp call a function in 1.lsp
4.  F12 on the function call in 2.lsp
5.  do the above step 2 times

I think after the step5 it should no longer call the parser to parse the 1.lsp file. And the 1.lsp atoms forest are cached.
But the codes is not that expected: https://github.com/Autodesk-AutoCAD/AutoLispExt/blob/f5d8e754065dafef1560cd034a2c77cb370fae49/extension/src/project/readOnlyDocument.ts#L75 create the ReadOnlyDocument each time, so there is no cache for atoms forest:
https://github.com/Autodesk-AutoCAD/AutoLispExt/blob/f5d8e754065dafef1560cd034a2c77cb370fae49/extension/src/project/readOnlyDocument.ts#L109

So it calls the https://github.com/Autodesk-AutoCAD/AutoLispExt/blob/f5d8e754065dafef1560cd034a2c77cb370fae49/extension/src/project/readOnlyDocument.ts#L112 every time.

And I tried to create a project file to contains above 1.lsp and 2.lsp, and only open 2.lsp file, it looks like this code:
https://github.com/Autodesk-AutoCAD/AutoLispExt/blob/f5d8e754065dafef1560cd034a2c77cb370fae49/extension/src/documents.ts#L57
The un-opened 1.lsp is also cached by Vscode document instead of the ReadOnlyDocument:
https://github.com/Autodesk-AutoCAD/AutoLispExt/blob/f5d8e754065dafef1560cd034a2c77cb370fae49/extension/src/documents.ts#L9

Reply:
- Your first scenario is dealing with documents that are Opened. We have no idea if there are unsaved edits and thus the only safe thing to do is regenerate from the TextDocument.
  - In the context of a rename operation it could be extremely destructive if we didn't regenerate.
  - If the document is opened I don't believe caching is a safe or feature complete option. Unsaved means we don't know if they've been working on it for the last 2 hours and just haven't saved.
- Your second scenario is because PRJ's have the opposite slash direction evenything else provides. Easy enough to normalize all of these; just noticed this the other day. Because they aren't 1-1 it never finds the existing ReadonlyDocument and keeps regenerating.

Reply:
So for the second scenario it looks like a bug?
And I just realized that the https://github.com/Autodesk-AutoCAD/AutoLispExt/blob/main/extension/src/documents.ts#L142 is only for `workspace` files.

Maybe in the class `DocumentManager` it should always cache the `ReadonlyDocument` for **all** files (workspace, and prj files, even the current file). When cache the ReadonlyDocument, it also caches the changed time stamp of this file; when to get the cache, try to check the changed time stamp, if it is out of date, it regennerates the atoms forest.  Do you think this is more cheaper way?

Reply:
On the ReadonlyDocument side, we have a direct fileContent property to read. So, I think the TextDocument.getText() on a max-range is cheap enough as long as we don't have resulting EOL comparison problems. 

I'll go make some more tweaks to recycle ROD's and fix the slashes, but I am going to leave PRJ isolated to the project for now. I really don't want the Project manager to be in constant contact with the Document manager if it can be avoided. Right now they have a very substantial separation of concerns and if we don't need to muddy those waters, then we really shouldn't.

Reply:
`but I am going to leave PRJ isolated to the project for now.`
I test that the above scenario is not an issue for slash. The two files are opened, so it always regenerate the atoms forest. I think after refining the ReadonlyDocument map, for each lisp file it only needs to regenerate atoms one time if it is not touched. (Also notice that if the opened files are also in project files, it will regenerate atoms twice: https://github.com/JD-Howard/AutoLispExt/blob/roDocSymbolMap/extension/src/providers/gotoProvider.ts#L53)